x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/ResultsIndexUpgradeService.java

```diff
@@ -0,0 +1,523 @@
+         */
+        RestStatus status = RestStatus.OK;
+        Throwable cause = null;
+        if (response.isTimedOut()) {
+            status = RestStatus.REQUEST_TIMEOUT;
+            cause = new ElasticsearchTimeoutException("Reindex request timed out");
+        }
+        for (BulkItemResponse.Failure failure : response.getBulkFailures()) {
+            if (failure.getStatus().getStatus() > status.getStatus()) {
+                status = failure.getStatus();
+                cause = failure.getCause();
+            }
+        }
+        for (ScrollableHitSource.SearchFailure failure : response.getSearchFailures()) {
+            RestStatus failureStatus = ExceptionsHelper.status(failure.getReason());
+            if (failureStatus.getStatus() > status.getStatus()) {
+                status = failureStatus;
+                cause = failure.getReason();
+            }
+        }
+        return new Tuple<>(status, cause);
+    }
+
+    /**
+     * Upgrade the indices given in the request.
+     *
+     * @param client   The client to use when making calls
+     * @param request  The upgrade request
+     * @param state    The current cluster state
+     * @param listener The listener to alert when actions have completed
+     */
+    public void upgrade(Client client, ResultsIndexUpgradeAction.Request request, ClusterState state,
+                        ActionListener<AcknowledgedResponse> listener) {
+        try {
+            checkMasterAndDataNodeVersion(state);
+            String[] concreteIndices = indexNameExpressionResolver.concreteIndexNames(state, request.indicesOptions(), request.indices());
+            MetaData metaData = state.getMetaData();
+
+            List<String> indicesToUpgrade = Arrays.stream(concreteIndices)
+                .filter(indexName -> shouldUpgrade.test(metaData.index(indexName)))
+                .collect(Collectors.toList());
+
+            // All the internal indices are up to date
+            if (indicesToUpgrade.isEmpty()) {
+                listener.onResponse(new AcknowledgedResponse(true));
+                return;
+            }
+
+            IndexNameAndAliasProvider indexNameAndAliasProvider = new IndexNameAndAliasProvider(indicesToUpgrade, metaData);
+            Exception validationException = indexNameAndAliasProvider.validate(metaData, shouldUpgrade);
+            if (validationException != null) {
+                listener.onFailure(validationException);
+                return;
+            }
+
+            // <7> Now that we have deleted the old indices, we are complete, alert the user
+            ActionListener<AcknowledgedResponse> deleteIndicesListener = ActionListener.wrap(
+                listener::onResponse,
+                error -> {
+                    logger.error("Failed to delete old indices: " + Strings.collectionToCommaDelimitedString(indicesToUpgrade),
+                        error);
+                    listener.onResponse(new AcknowledgedResponse(true));
```

Comment:
why not `listener.onFailure`?

Reply:
Yeah, now that I think about it more, a new job that is created that needs to use the shared results index would be using the OLD index that was not deleted appropriately. This would be problematic. We should fail here.